File: day12/solve.ts
import { solve } from "../runner/typescript";
import { max, sum, groupBy } from "lodash";

type Row = readonly [string, readonly number[]];

function parser(input: string) {
  const lines = input.split("\n");
  const result = lines.map((line: string) => {
    const [springs, groups] = line.split(" ");
    return [springs, groups.split(",").map(Number)] as const;
  });
  return result;
}

const test = countArrangements(["?###????????", [3, 2, 1]]);
console.log(test);

function isValid(springs: string, groups: readonly number[]): boolean {
  const total = springs.split("").filter((s) => s === "#").length;
  if (total !== sum(groups)) return false;

  // groups of coninuous springs
  let currentGroupLen = 0;
  let currentGroupIndex = 0;
  for (const spring of springs.split("")) {
    if (spring === "#") {
      currentGroupLen++;
      continue;
    }

    if (currentGroupLen === 0) continue;
    if (currentGroupLen !== groups[currentGroupIndex]) {
      return false;
    }
    currentGroupLen = 0;
    currentGroupIndex += 1;
  }
  if (currentGroupLen === 0 || currentGroupLen === groups[currentGroupIndex]) {
    return true;
  }
  return false;
}

function countArrangements([springs, groups]: Row): number {
  const maxDamanged = sum(groups);
  const maxOperational = springs.length - maxDamanged;
  const found = [];
  const queue = [springs];
  while (queue.length > 0) {
    const current = queue.shift();

    const countDamaged = current.split("#").length - 1;
    if (countDamaged > maxDamanged) continue;
    const countOperational = current.split(".").length - 1;
    if (countOperational > maxOperational) continue;

    if (current.includes("?")) {
      queue.push(current.replace("?", "#"));
      queue.push(current.replace("?", "."));
    } else {
      if (isValid(current, groups)) found.push(current);
    }
  }
  console.log(found.length);
  return found.length;
}

function part1(rows: Row[]): number {
  const totals = rows.map(countArrangements);
  console.log({ totals });
  return sum(totals);
}

function part2(rows: Row[]): number {
  const totals = rows
    .map(
      ([springs, groups]) =>
        [
          `${springs}?${springs}?${springs}?${springs}?${springs}`,
          [...groups, ...groups, ...groups, ...groups, ...groups],
        ] as const
    )
    .map(countArrangements);
  console.log({ totals });
  return sum(totals);
}

const testInput =
  "???.### 1,1,3\n.??..??...?##. 1,1,3\n?#?#?#?#?#?#?#? 1,3,1,6\n????.#...#... 4,1,1\n????.######..#####. 1,6,5\n?###???????? 3,2,1";
solve({
  parser: parser,
  part1: part1,
  // part2: part2,

  part1Tests: [[testInput, 21]],
  part2Tests: [[testInput, 525152]],
});

File: day12/solve.py
from runner.python import solve
from functools import cache


def parse(input: str) -> list[tuple[str, tuple[int, ...]]]:
  return [
    (line.split()[0], tuple(map(int, line.split()[1].split(","))))
    for line in input.splitlines()
  ]


@cache
def count_arrangements(springs: str, groups: tuple[int, ...], current_group_length=0):
  # When at end of given groups, each group after is 0 length
  group = groups[0] if groups else 0

  if not springs:  # base case - end of possible spring string
    if len(groups) > 1 or group != current_group_length:
      return 0  # leftover groups or leftover group length
    return 1

  # parse each char of the string recursively
  spring, springs = springs[0], springs[1:]
  match spring:
    case "?":
      damaged = count_arrangements("#" + springs, groups, current_group_length)
      operational = count_arrangements("." + springs, groups, current_group_length)
      return damaged + operational

    case "#":
      if current_group_length > group:  # group too long
        return 0

      return count_arrangements(springs, groups, current_group_length + 1)

    case ".":
      if current_group_length:  # end of group
        if current_group_length == group:
          return count_arrangements(springs, tuple(groups[1:]), 0)
        return 0

      return count_arrangements(springs, groups, 0)


def part1(input):
  counts = [count_arrangements(*row) for row in parse(input)]
  return sum(counts)


def part2(input):
  unfolded = [("?".join([s] * 5), g * 5) for s, g in parse(input)]
  counts = [count_arrangements(*row) for row in unfolded]
  return sum(counts)


testInput = "???.### 1,1,3\n.??..??...?##. 1,1,3\n?#?#?#?#?#?#?#? 1,3,1,6\n????.#...#... 4,1,1\n????.######..#####. 1,6,5\n?###???????? 3,2,1"
solve(
  part1,
  testInput,
  21,
  part2,
  testInput,
  525152,
)
